### PR TITLE
add IntoFuture trait and support for await

### PR DESCRIPTION
**Pull request by seanmonstar**
The [async-await RFC](https://rust-lang.github.io/rfcs/2394-async_await.html#the-await-compiler-built-in) mentions being able to `await` anything implementing `IntoFuture`. Somewhere along the way, it was left out.

### CODE REVIEW

**Comment by seanmonstar** [comment]
(My machine dies trying to compile rustc, so I'll be needing CI to tell me if I goofed a small thing. And I can add a tracking issue if accepted.)

File: src/librustc/hir/lowering/expr.rs
Comment (Centril) [inline src/librustc/hir/lowering/expr.rs]: This `let` binding feels rather redundant; let's inline it into where it is used (and inline the tail expression into the `let into_future_expr`).

Comment (seanmonstar) [reply src/librustc/hir/lowering/expr.rs]: I copied the pattern from the `IntoIterator` code ;)

File: src/librustc/hir/lowering/expr.rs
Comment (Centril) [inline src/librustc/hir/lowering/expr.rs]: This won't work, I think... you need to mark the span as allowing the feature with https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/lowering/struct.LoweringContext.html#method.mark_span_with_reason.

File: src/librustc/hir/lowering/expr.rs
Comment (Centril) [reply src/librustc/hir/lowering/expr.rs]: Figured, let's fix the `IntoIterator` code also then.

File: src/librustc/hir/lowering/expr.rs
Comment (Centril) [inline src/librustc/hir/lowering/expr.rs]:
```suggestion
        let expr = self.lower_expr(expr);
```

File: src/test/ui/async-await/await-into-future.rs
Comment (nikomatsakis) [inline src/test/ui/async-await/await-into-future.rs]: Doesn't this need to be imported? Also, I'd expect some stability annotations. Perhaps the tests will fail, though. Otherwise, seems like we have a "hole" :)

File: src/libcore/future/future.rs
Comment (nikomatsakis) [inline src/libcore/future/future.rs]: We should file a tracking issue, but let's hold off just a bit.

**Review by nikomatsakis — DISMISSED** [review]
The code generally looks good to me -- though I feel like the test ought not to be passing. Let's see what travis says. 

Anyway, the other question is whether we want to enable this. It'd be good to capture some of the motivation. The main thing I've heard (which I find persuasive) came from [the async-finalizers blog post](https://blog.yoshuawuyts.com/async-finalizers/) by @yoshuawuyts. The main reason I've heard for *removing* `IntoFuture` was basically "YAGNI" -- not sure if there were more motivations than that? (Maybe @withoutboats remembers?)

**Comment by cramertj** [comment]
I don't think it was forgotten about-- `IntoFuture` was intentionally removed from futures-rs a while back because the only things left that implemented it were tuple types. I'm not opposed to this PR at all; I don't know of any serious downsides to providing this functionality, aside from potentially yet another layer of indirection for error messages. However, I am curious to see what use-cases you have in mind for it!

**Comment by nikomatsakis** [comment]
@cramertj did you see [the async-finalizers (since renamed to async-builders, I think) blog post](https://blog.yoshuawuyts.com/async-finalizers/) I was referring to?  That's at least one use case.

(I'd like to know if @seanmonstar had another use case in mind, though!)

**Comment by seanmonstar** [comment]
Yes, I've long had various builders in reqwest and hyper that I've previously felt would be nice if they could easily become futures, but they themselves aren't futures.

(My office internet is flaking hard, so might take a bit to update the PR.)

File: src/librustc/hir/lowering/expr.rs
Comment (seanmonstar) [reply src/librustc/hir/lowering/expr.rs]: Updated (only here, left `IntoIterator` alone in this PR).

File: src/librustc/hir/lowering/expr.rs
Comment (seanmonstar) [reply src/librustc/hir/lowering/expr.rs]: Oh cool! Got it, I think?

**Comment by seanmonstar** [comment]
It *appears* everything else worked, so I've kicked off a local build to `--bless` the new output. Hopefully it doesn't get OOMed...

**Comment by seanmonstar** [comment]
Nevermind, it seems there's codegen issues that I don't yet understand:

- ` ui/async-await/async-fn-size.rs` is type overflowing `Into::into`. My guess is it's coming from `Box::pin(fut)`, but not sure why yet.
- `ui/async-await/async-fn-size-moved-locals.rs` has noticed a `mem::size_of_val` change of 4 bytes on `joined`. I suppose this change adds a new local that impact layout?

**Comment by Centril** [comment]
> `ui/async-await/async-fn-size-moved-locals.rs` has noticed a `mem::size_of_val` change of 4 bytes on `joined`. I suppose this change adds a new local that impact layout?

cc @tmandry

File: src/librustc/hir/lowering/expr.rs
Comment (alex) [inline src/librustc/hir/lowering/expr.rs]:
```suggestion
        // `match ::std::future::IntoFuture::into_future(<expr>) { ... }`
```

**Comment by tmandry** [comment]
> * `ui/async-await/async-fn-size-moved-locals.rs` has noticed a `mem::size_of_val` change of 4 bytes on `joined`. I suppose this change adds a new local that impact layout?

I can look into what's going on, but a single shift by 4 bytes isn't a huge deal imo. I still want to make the layout algorithm smarter and sometimes it does dumb things (see #62575). #61430 also made some sizes worse.

**Comment by withoutboats** [comment]
I'm in favor of this change! In a library I've been working on the last few weeks I've found that I wanted an into_future support in await as well. As @cramertj said, the only downside could be from an impact on diagnostics - we should make sure error messages don't regress at all as a part of this change.

However, I'm not sure procedurally what we should do. I think we definitely need an FCP on adding this to nightly, but I don't know if an additional RFC is necessary.

**Comment by Centril** [comment]
[We are also discussing this PR on Zulip.](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/.2365244.20-.20IntoFuture)

**Comment by seanmonstar** [comment]
So, the rest of things are fixed, the one last outstanding failing test is that this change is causing the generator futures in `src/test/ui/async-await/async-fn-size.rs` to hit the "type-length limit". I don't know if this can be avoided, or what. (It compiles if I add `#![type_length_limit="3166948"]` to that test file XD).

File: src/libcore/future/future.rs
Comment (yoshuawuyts) [inline src/libcore/future/future.rs]: [`std::iter::IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) has an associated `Item` type:

```rust
pub trait IntoIterator {
    type Item;
    type IntoIter: Iterator;
    fn into_iter(self) -> Self::IntoIter;
}
```

I'm thinking it would make sense for `IntoFuture` to be consistent with it and similarly define an `Output` type:

```rust
pub trait IntoFuture {
    type Output;
    type Future: Future<Output = Self::Output>;
    fn into_future(self) -> Self::Future;
}
```

It's not a big deal, but having consistency between similar traits generally seems desirable. This is also why we've chosen that direction for async-std's [`IntoFuture`](https://github.com/async-rs/async-std/pull/259) and [`IntoStream`](https://docs.rs/async-std/0.99.9/async_std/stream/trait.IntoStream.html) APIs.

Comment (withoutboats) [reply src/libcore/future/future.rs]: The reason for this is that it can help avoid double indirection in associated types, `<<T as IntoFuture>::Future as Future>::Output` is quite gnarly, and its very nice to be able to write `<T as IntoFuture>::Output` instead.

File: src/test/ui/async-await/async-fn-nonsend.stderr
Comment (nikomatsakis) [inline src/test/ui/async-await/async-fn-nonsend.stderr]: My opinion is that these diagnostics, while worse than the originals, were already pretty bad, so I don't think it's a problem to land them as is.

**Comment by nikomatsakis** [comment]
We concluded in our lang team meeting on [2019-10-10](https://github.com/rust-lang/lang-team/blob/master/minutes/2019-10-10.md) that we would handle this as follows:

* Somebody write up a brief motivation showing the use cases 
    * this is probably the builder pattern that @yoshuawuyts described, so this could be cribbed from their blog post
* We will then rfcbot fcp merge and land this on nightly
* We can collect any add'l use cases or comments
* And stabilize with a suitable stabilization report

File: src/libcore/future/future.rs
Comment (nikomatsakis) [reply src/libcore/future/future.rs]: I agree with this proposal.

Comment (seanmonstar) [reply src/libcore/future/future.rs]: Done.

File: src/test/ui/async-await/async-fn-nonsend.stderr
Comment (Centril) [reply src/test/ui/async-await/async-fn-nonsend.stderr]: @estebank This does remind me however of the issue that we need to shorten paths in diagnostics for paths in scope... when will that be fixed? :D

Comment (estebank) [reply src/test/ui/async-await/async-fn-nonsend.stderr]: I think that further progress along the lines of https://github.com/rust-lang/rust/pull/65456 might help here. I tried rebasing this PR on top of it, but when running this test I get the following with no span

    error[E0433]: failed to resolve: could not find `IntoFuture` in `future`

**Comment by tmandry** [comment]
As far as I can tell, future size is no longer a concern for this PR (there are a couple examples that get smaller, none get larger).

We should have an issue to track improvements to the diagnostic, though.

Increasing the `type_length_limit` is probably unavoidable. Nested futures tend to produce large types. (If we think there's a way to improve that situation, there should also be an issue to track that.)

r=me with passing tests and tracking issue filed.

**Comment by seanmonstar** [comment]
> Increasing the type_length_limit is probably unavoidable.

My one concern is that async code that compiled before (1.39) may suddenly break when a user upgrades their compiler. If there's nothing that can be done, I can up the limit on that test and call it a day...

**Comment by cramertj** [comment]
> My one concern is that async code that compiled before (1.39) may suddenly break when a user upgrades their compiler. If there's nothing that can be done, I can up the limit on that test and call it a day...

I think this unfortunately is in about the same state as changing a `macro_rules` macro to one that recurses more, forcing the user to raise the recursion limit.

**Comment by seanmonstar** [comment]
CI is green. So, shall I just create a tracking issue for this trait and update? Or is more discussion still needed?

**Comment by tmandry** [comment]
I see. Yes, I think this warrants more discussion first.

It would be nice if we could backport and make that problem go away, but that's really hard to justify.

One idea I had was to just "artificially lower the limit for Future types," backporting to give us a "budget" to spend later. But it doesn't seem like we can easily do that. Prior to this change, in this test there's simply no function type which gets monomorphized with the full future type embedded. We would need to to "simulate" this monomorphization of `into_future()` somehow for every `.await` point. I'm not exactly sure how to do that, or that it would be a small enough change to backport.

Finally, without backporting anything, we might be able to special-case the type length check for `into_future` specifically. But I still don't know how we would implement that without hanging on enormous types, and then we'd have to keep this special casing around forever.

cc @arielb1 @eddyb for more discussion. Maybe we can come up with a better way. If not, are people comfortable with merging this?

**Comment by tmandry** [comment]
I think my preferred solution is to extract the type length limit check as a helper function, and perform it at some point (??) when a `.await` point is encountered. (Assuming this is actually feasible, and we can find a proper place to add this check, and the change is small enough to comfortably backport.)

NB: We can add to this length to "simulate" the type length of a monomorphized `into_future()` function, and check against that length. But 99% of the time this won't matter, since the main difference is that we perform the type length check when we didn't before.

**Comment by eddyb** [comment]
`type_length_limit` is a codegen thing (so I doubt `.await` would make sense with it) and it has some issues we need to fix, but I can't tell if they're relevant here. cc @pnkfelix

File: src/librustc/hir/lowering/expr.rs
Comment (eddyb) [inline src/librustc/hir/lowering/expr.rs]: Oh, I think I know why that test changed, but also, that it's entirely avoidable.
What seems to be happening is something very silly which is that if `<expr>` has any `.await`s itself, the `IntoFuture:: into_future` value is considered to be "live at a suspend point".

I don't see why MIR building would generate that instead of keeping a constant operand around, but if nobody has the time to fix this for all `f(g().await)` (i.e. make it equivalent to `{let x = g().await; f(x)}`) you can do a manual transformation here and leave a FIXME comment pointing to an issue to clean it up.
(cc @matthewjasper @Zoxc for the MIR side of this)

The manually transformed form would be something like `match { let x = <expr>; IntoFuture:: into_future(x) } {...}`.

Comment (eddyb) [reply src/librustc/hir/lowering/expr.rs]: Just remembered that typeck collects a list of types without being aware of whether they might show up in the MIR.

I think we can whitelist call expressions where the callee is a path to a function or method (not based on type, since you can put a function in a local variable).

**Comment by tmandry** [comment]
> `type_length_limit` is a codegen thing (so I doubt `.await` would make sense with it)

I don't see anything codegen-specific in the implementation:

https://github.com/rust-lang/rust/blob/4a8c5b20c7772bc5342b83d4b0696ea216ef75a7/src/librustc_mir/monomorphize/collector.rs#L477

To be clear, I was suggesting we replicate this check earlier on in the compiler, as a sort of "forward-compatibility hack" for this change. This is not great, but we would be able to remove this hack eventually.

That said, your fix would be strongly preferable to this one.

**Comment by eddyb** [comment]
Note the path: `rustc_mir::monomorphize::collector` - that collects monomorphizations for codegen.
I think the type length limit might have something to do with very long LLVM type names?
Ideally it wouldn't matter if human-readable LLVM IR is not being emitted.

We can probably get rid of that limit without negative consequence, although it shouldn't matter either way here since the type *really* shouldn't have changed.

**Comment by Centril** [comment]
I would strongly prefer not to introduce async-specific hacks to the type length limit at any rate; if we can do *general* and principled improvements then great.

File: src/librustc/hir/lowering/expr.rs
Comment (seanmonstar) [reply src/librustc/hir/lowering/expr.rs]: > The manually transformed form would be something like `match { let x = <expr>; IntoFuture:: into_future(x) } {...}`.

I don't grasp the *why*, but should I try this? Local testing is hard, but I can make this change and remove the `type_level_limit` attribute from that test, push and see if CI passes.

Comment (eddyb) [reply src/librustc/hir/lowering/expr.rs]: The why is easy: `f(g().await)` right now keeps a temporary `f` around across the suspend point for the inner await - or at least `rustc_typeck` assumes that happens.

It might help if you think of this as `call(f, call(g).await)`, where `f` is necessarily evaluated *before* the suspend, but the outer `call` is necessarily evaluated *after* the suspend, requiring the result of evaluating `f` to stick around (which also means its type ends up in the generator type).

**Comment by tmandry** [comment]
> I would strongly prefer not to introduce async-specific hacks to the type length limit at any rate; if we can do _general_ and principled improvements then great.

Agreed, but if we couldn't find a more general solution, I'd still prefer a temporary hack to breaking people's working code when they upgrade the compiler.

File: src/librustc/hir/lowering/expr.rs
Comment (tmandry) [reply src/librustc/hir/lowering/expr.rs]: The MIR should be irrelevant since the type we're instantiating is decided by typeck. That information is in the `GeneratorInterior`, built by the code in this file: https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/generator_interior.rs#L169

I'm not sure exactly what special case @eddyb had in mind, but I think it would go in one of the methods of the `InteriorVisitor`.

**Comment by Centril** [comment]
> a temporary hack

These tend to become permanent. ;)

File: src/librustc/hir/lowering/expr.rs
Comment (eddyb) [reply src/librustc/hir/lowering/expr.rs]: Hopefully we can rely on https://github.com/rust-lang/rust/pull/65743 for the fix.

**Comment by eddyb** [comment]
I've opened a PR for the approach I suggested: https://github.com/rust-lang/rust/pull/65743.

**Comment by tmandry** [comment]
> These tend to become permanent. ;)

I only suggested it because it could be removed as soon as (a) this change got merged or (b) we decided we weren't going to adopt the change. (I should have been more explicit about that, but at this point I don't think it's relevant =)

**Comment by seanmonstar** [comment]
Besides the slightly changed diagnostics, it seems using the `match { let x = expr; IntoFuture::into_future(x) } { ... }` doesn't fix the `async-fn-size` issue about `type_level_limit`.

@eddyb claims that rebasing this on his fix in #65743 *does* fix the `type_level_limit`, so I can rebase when that is merged.

**Comment by danieleades** [comment]
pretty excited for this. any further progress?

**Comment by seanmonstar** [comment]
I've rebased on master such that it includes @eddyb's fix, and in doing so, removed the `type_level_limit` change in the test. It's all passing now.

Does this still need an FCP before I open a tracking issue and update here?

File: src/libcore/future/future.rs
Comment (tmandry) [reply src/libcore/future/future.rs]: There should be a tracking issue before this gets merged, though, I think.

File: src/librustc/hir/lowering/expr.rs
Comment (tmandry) [inline src/librustc/hir/lowering/expr.rs]: this comment should be fixed

**Comment by tmandry** [comment]
@bors delegate=seanmonstar

r=me with a tracking issue opened and conflicts fixed.

**Comment by seanmonstar** [comment]
@tmandry I hadn't done the tracking issue stuff yet since it seemed this needed an FCP. It's still tagged as such...

**Comment by seanmonstar** [comment]
@Centril as you added the `needs-fcp` label, how do you feel? Time to trigger it?

**Comment by Centril** [comment]
Let's do it. :) Dear language team, lib team, and the community; I propose that we experimentally try out `IntoFuture` in the lowering, with more use cases to be collected during experimentation, and then eventually stabilized if all goes well.

@rfcbot merge

**Comment by sfackler** [comment]
Ping @Kimundi @KodrAus @SimonSapin @joshtriplett @nikomatsakis @pnkfelix @scottmcm @withoutboats

File: src/libcore/future/future.rs
Comment (pnkfelix) [inline src/libcore/future/future.rs]: Is the `IntoFuture::Output` associated type just a convenience? That is, can't we leave this out, and have people "just" write `IntoFuture::Future::Output` instead?

Comment (withoutboats) [reply src/libcore/future/future.rs]: It is, but its consistent with the API for IntoIterator.

Comment (sfackler) [reply src/libcore/future/future.rs]: You also can't write `IntoFuture::Future::Output` - it has to be `<IntoFuture::Future as Future>::Output` D:

Comment (yoshuawuyts) [reply src/libcore/future/future.rs]: Prior discussion about this: https://github.com/rust-lang/rust/pull/65244#pullrequestreview-301916859

**Comment by danieleades** [comment]
The suspense is killing me

**Comment by cramertj** [comment]
Ping @Kimundi @joshtriplett @scottmcm

**Comment by joshtriplett** [comment]
Question on intent: would people expect await on a tuple or collection of futures to await them in parallel, rather than in series?

If so, this change seems likely to significantly improve ergonomics for correct usage of futures.

**Comment by jplatte** [comment]
I wouldn't expect await on a tuple or collection of futures towwork at all and even if it worked I think I'd rather pass that to a helper function that makes it clear in the code how the futures are awaited. `(fut1, fut2, fut3).await` is a bit too magical for me.

**Comment by Centril** [comment]
While I think `IntoFuture` is a good idea, I'm not in favor of tuples of `Future`s or `IntoFuture`s being `.await`able because:

- I don't think that `(a, b, c).await` vs. `join!(a, b, c)` has a significant difference wrt. ergonomics.

- There's a semantic choice: should it sequentially `.await` or should it be like `join!(...)`. I think this choice is not obvious and so while it might be a convenience when writing code, I think explicitly writing out `join!(...)` (or better yet, `(a, b, c).join!()`) is more understandable as opposed to trying to design a DWIM feature.

   - This is analogous to how we refuse to make a choice wrt. tuples of iterators (Zipping or Cartesian product?).

- Similarly, we do not e.g. allow a tuple of closures, e.g. `(f, g, h)` to be callable, e.g. `(f, g, h)()`.

**Comment by joshtriplett** [comment]
What would be the argument for awaiting sequentially rather than joining? If you have multiple futures, why *not* await in parallel?

I expect one of the most common *problems* with futures to be the unnecessarily sequential awaiting of futures.

Awaiting a tuple doesn't seem magical to me, it seems intuitive.

**Comment by sfackler** [comment]
I don't think consensus on the question of awaiting on tuples needs to block consensus on the addition of IntoFuture.

**Comment by Centril** [comment]
> What would be the argument for awaiting sequentially rather than joining? If you have multiple futures, why _not_ await in parallel?

The user might want deterministic ordering for the effects that `.await`ing has and there might not be data-dependence (but there might be effect-dependence) which would force first doing `let a = fa.await; let b = fb(a).await;`.

I think `join!(...)` clarifies the semantics well without being any longer.

> I don't think consensus on the question of awaiting on tuples needs to block consensus on the addition of IntoFuture.

Yeah, I agree with this.

**Comment by joshtriplett** [comment]
I'm not looking to block on that, I was looking to evaluate potential use cases and semantics.

**Comment by sfackler** [comment]
Cool - can you tick your checkbox then?

**Comment by nikomatsakis** [comment]
I agree that we should defer a discussion of `await` on tuples -- it needn't block this change, which I think is well-motivated either way.

**Comment by cramertj** [comment]
This PR has been waiting on checkboxes for more than a month now despite no objection to the original proposal. Ping @joshtriplett @scottmcm @Kimundi for checkboxes again unless you have remaining concerns.

**Comment by dtolnay** [comment]
I have checked Kimundi's box, see https://github.com/rust-lang/team/pull/200.

**Pull request merged**
